From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Tue, 8 Mar 2016 20:47:08 +0100
Subject: [PATCH] {BUG} {WIP} vmxnet3: identify unbalanced locking through
 uninitialised var

clang reports:

  CC [M]  drivers/net/vmxnet3/vmxnet3_drv.o
drivers/net/vmxnet3/vmxnet3_drv.c:1029:9: error: variable 'flags' is used uninitialized whenever 'if' condition is
      true [-Werror,-Wsometimes-uninitialized]
                                if (unlikely(ctx.eth_ip_hdr_size +
                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
include/linux/compiler.h:166:22: note: expanded from macro 'unlikely'
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
drivers/net/vmxnet3/vmxnet3_drv.c:1129:39: note: uninitialized use occurs here
        spin_unlock_irqrestore(&tq->tx_lock, flags);
                                             ^~~~~
drivers/net/vmxnet3/vmxnet3_drv.c:1029:5: note: remove the 'if' if its condition is always false
                                if (unlikely(ctx.eth_ip_hdr_size +
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
drivers/net/vmxnet3/vmxnet3_drv.c:1023:8: error: variable 'flags' is used uninitialized whenever 'if' condition is
      true [-Werror,-Wsometimes-uninitialized]
                        if (unlikely(ctx.eth_ip_hdr_size + ctx.l4_hdr_size >
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
include/linux/compiler.h:166:22: note: expanded from macro 'unlikely'
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
drivers/net/vmxnet3/vmxnet3_drv.c:1129:39: note: uninitialized use occurs here
        spin_unlock_irqrestore(&tq->tx_lock, flags);
                                             ^~~~~
drivers/net/vmxnet3/vmxnet3_drv.c:1023:4: note: remove the 'if' if its condition is always false
                        if (unlikely(ctx.eth_ip_hdr_size + ctx.l4_hdr_size >
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
drivers/net/vmxnet3/vmxnet3_drv.c:977:21: note: initialize the variable 'flags' to silence this warning
        unsigned long flags;
                           ^
                            = 0
2 errors generated.

This is a spin lock is unlocked without being locked first.
---
 drivers/net/vmxnet3/vmxnet3_drv.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/vmxnet3/vmxnet3_drv.c b/drivers/net/vmxnet3/vmxnet3_drv.c
index fc895d0e85d9..a07ed37ab5cb 100644
--- a/drivers/net/vmxnet3/vmxnet3_drv.c
+++ b/drivers/net/vmxnet3/vmxnet3_drv.c
@@ -974,7 +974,7 @@ vmxnet3_tq_xmit(struct sk_buff *skb, struct vmxnet3_tx_queue *tq,
 {
 	int ret;
 	u32 count;
-	unsigned long flags;
+	unsigned long flags = 0 /* BUG > v4.5-rc7 : goto hdr_too_big without spin_lock_irqsave */;
 	struct vmxnet3_tx_ctx ctx;
 	union Vmxnet3_GenericDesc *gdesc;
 #ifdef __BIG_ENDIAN_BITFIELD
-- 
